pkg/apiserver/proxy_test.go

```diff
@@ -34,6 +36,115 @@ import (
 	"k8s.io/kubernetes/pkg/api/rest"
 )
 
+func calculateContentLength(reqBody string, transferEncodings []string) int64 {
+	if len(transferEncodings) == 0 || (len(transferEncodings) == 1 && transferEncodings[0] == "identity") {
+		return int64(len(reqBody))
+	}
+	// RFC2616 section 4.4: The Content-Length header field MUST NOT be sent if these two lengths
+	// are different (i.e., if a Transfer-Encoding header field is present).
+	return int64(-1)
+}
+
+func containGzip(transferEncodings []string) bool {
+	for _, encoding := range transferEncodings {
+		if encoding == "gzip" {
+			return true
+		}
+	}
+	return false
+}
+
+func TestProxyRequestContentLengthAndTransferEncoding(t *testing.T) {
+	serverResponse := "got response"
+	table := []struct {
+		transferEncodings []string
+		reqBody           string
+		reqNamespace      string
+	}{
+		{[]string{}, "", "default"},
+		{[]string{"identity"}, "question", "default"},
+		{[]string{"chunked"}, "question", "default"},
+		// RFC2616 section-3.6: Whenever a transfer-coding is applied to a message-body, the set of
+		// transfer-codings MUST include "chunked", unless the message is terminated by closing the
+		// connection.
+		{[]string{"chunked", "gzip"}, "qqqqqqqqqquestion", "default"},
+	}
+
+	for _, item := range table {
+		proxyServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
```

Comment:
Suggest naming this "downstreamServer"-- it's not the proxy, it's the thing the proxy dials, right?